libsolidity/formal/SymbolicState.cpp

```diff
@@ -149,6 +153,120 @@ smtutil::Expression SymbolicState::txConstraints(FunctionDefinition const& _func
 	return conj;
 }
 
+void SymbolicState::buildABIFunctions(set<FunctionCall const*> const& _abiFunctions)
+{
+	map<string, SortPointer> functions;
+
+	for (auto const* funCall: _abiFunctions)
+	{
+		auto t = dynamic_cast<FunctionType const*>(funCall->expression().annotation().type);
+
+		auto const& args = funCall->sortedArguments();
+		auto const& paramTypes = t->parameterTypes();
+		auto const& returnTypes = t->returnParameterTypes();
+
+
+		auto argTypes = [](auto const& _args) {
+			return applyMap(_args, [](auto arg) { return arg->annotation().type; });
+		};
+
+		vector<TypePointer> inTypes;
+		vector<TypePointer> outTypes;
+		if (t->kind() == FunctionType::Kind::ABIDecode)
+		{
+			inTypes.emplace_back(TypeProvider::bytesMemory());
+			auto const* tupleType = dynamic_cast<TupleType const*>(args.at(1)->annotation().type);
+			solAssert(tupleType, "");
+			for (auto t: tupleType->components())
+			{
+				auto typeType = dynamic_cast<TypeType const*>(t);
+				solAssert(typeType, "");
+				outTypes.emplace_back(typeType->actualType());
+			}
+		}
+		else
+		{
+			outTypes = returnTypes;
+			if (
+				t->kind() == FunctionType::Kind::ABIEncodeWithSelector ||
+				t->kind() == FunctionType::Kind::ABIEncodeWithSignature
+			)
+			{
+				inTypes.emplace_back(paramTypes.front());
+				inTypes += argTypes(vector<ASTPointer<Expression const>>(args.begin() + 1, args.end()));
+			}
+			else
+			{
+				solAssert(
+					t->kind() == FunctionType::Kind::ABIEncode ||
+					t->kind() == FunctionType::Kind::ABIEncodePacked,
+					""
+				);
+				inTypes = argTypes(args);
+			}
+		}
+
+		auto replaceTypes = [](auto& _types) {
```

Comment:
So I guess here we could bring in ``encodingType``s in a later step, right? The difficulty being to determine if a type and its encoding type are both actually mapped to the same SMT sorts, s.t. the resulting function remains type-correct when applying to the SMT expressions later?

Reply:
Yea, we could maybe even do that in this PR

Reply:
As you like - it'd be a strict refinement and a rather local change probably, so also no problem doing it later.